docs/fundamentals/code-analysis/quality-rules/ca1839.md

```diff
@@ -0,0 +1,57 @@
+---
+title: "CA1839: Use `string.Contains(char)` instead of `string.Contains(string)` with single characters"
+description: "Consider using `string.Contains(char)` instead of `string.Contains(string)` when searching for a single character"
+ms.date: 03/03/2021
+ms.topic: reference
+f1_keywords:
+  - "CA1839"
+helpviewer_keywords:
+  - "CA1839"
+author: MeikTranel
+---
+
+# CA1839: Use `string.Contains(char)` instead of `string.Contains(string)` with single characters
+
+| | Value |
+|-|-|
+| **Rule ID** |CA1839|
+| **Category** |[Performance](performance-warnings.md)|
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+`string.Contains(string)` is used when `string.Contains(char)` was available.
+
+## Rule description
+
+When searching for a single character using `string.Contains(char)` offers better performance than `string.Contains(string)`.
+
+## How to fix violations
+
+In general, the rule is fixed simply by using a char literal instead of a string literal.
+
+```csharp
+public bool ContainsLetterI()
+{
+    var testString = "I am a test string.";
+    return testString.Contains("I");
+}
+```
```

Comment:
Can you add the vb equivalent and dev_langs metadata?

Reply:
Of course - question is if I have to use the file reference syntax into the snippets folder or if there is a way to side-by-side the code examples inline of the markdown file since the snippet is so short.

Reply:
@MeikTranel The benefit of a snippet file is having build verification for the code in it.

I'd wait for @gewarren to tell if you have to use it or it's okay to keep the code in markdown.

Reply:
When it's this short I'm okay with putting it inline. You can just stack the snippets in the markdown file. Only one will be displayed depending on the user's selected language.

Reply:
I would do something like this:

```cs
public bool ContainsLetterI()
{
    var testString = "I am a test string.";
    return testString.Contains("I"); // violation of CA1839
}

// solution for .NET Standard 2.1, .NET Core 3.1, and .NET 5 or newer: 
public bool ContainsLetterI()
{
    var testString = "I am a test string.";
    return testString.Contains('I'); // fix for CA1839.
}

// fix for any Target Framework older than .NET Standard 2.1, and .NET Core 3.1 (such as targeting .NET Framework for example).
public bool ContainsLetterI()
{
    var testString = "I am a test string.";
    return testString.IndexOf('I') > -1; // fix for CA1839.
}
```

```vb
Public Function ContainsLetterI as Boolean
    Dim testString As String = "I am a test string."
    Return testString.Contains("I") ' violation of CA1839
End Function

' solution for .NET Standard 2.1, .NET Core 3.1, and .NET 5 or newer:
Public Function ContainsLetterI as Boolean
    Dim testString As String = "I am a test string."
    Return testString.Contains("I"c) ' fix for CA1839.
End Function


// fix for any Target Framework older than .NET Standard 2.1, and .NET Core 3.1 (such as targeting .NET Framework for example).
Public Function ContainsLetterI as Boolean
    Dim testString As String = "I am a test string."
    Return testString.IndexOf("I"c) > -1 ' fix for CA1839.
End Function
```

Also I would recommend trying to make an F# version as well too for those who use F#.

Reply:
@AraHaan I don't think CAxxxx rules are applicable to F#

Reply:
Might not but only one way to find out by testing that.

Reply:
What do you mean "by testing that"? Analyzers have applicable languages as seen in the other PR you commented on - this is a "roslyn" analyzers rule - last i checked fsharp doesnt use roslyn. If you want this to be supported in fsharp - please ask the fsharp people for support.
As for the requested change i would recommend you open a separate issue for this. This is guidance that we don't offer anywhere else and the underlying issue is definitely not unique to this rule. Its a simple rule and the documentation should be simple as well imho.